source/opt/debug_info_manager.cpp

```diff
@@ -210,6 +212,52 @@ uint32_t DebugInfoManager::BuildDebugInlinedAtChain(
   return chain_head_id;
 }
 
+Instruction* DebugInfoManager::GetDebugOperationWithDeref() {
+  if (deref_operation_ != nullptr) return deref_operation_;
+
+  uint32_t result_id = context()->TakeNextId();
+  std::unique_ptr<Instruction> deref_operation(new Instruction(
+      context(), SpvOpExtInst, context()->get_type_mgr()->GetVoidTypeId(),
+      result_id,
+      {
+          {SPV_OPERAND_TYPE_ID,
+           {context()
+                ->get_feature_mgr()
+                ->GetExtInstImportId_OpenCL100DebugInfo()}},
+          {SPV_OPERAND_TYPE_EXTENSION_INSTRUCTION_NUMBER,
+           {static_cast<uint32_t>(OpenCLDebugInfo100DebugOperation)}},
+          {SPV_OPERAND_TYPE_CLDEBUG100_DEBUG_OPERATION,
+           {static_cast<uint32_t>(OpenCLDebugInfo100Deref)}},
+      }));
+
+  // Add to the front of |ext_inst_debuginfo_|.
+  deref_operation_ =
+      context()->module()->ext_inst_debuginfo_begin()->InsertBefore(
+          std::move(deref_operation));
+
+  RegisterDbgInst(deref_operation_);
+  if (context()->AreAnalysesValid(IRContext::Analysis::kAnalysisDefUse))
+    context()->get_def_use_mgr()->AnalyzeInstDefUse(deref_operation_);
+  return deref_operation_;
+}
+
+Instruction* DebugInfoManager::CloneDebugExpressionAndPushDerefOperation(
+    Instruction* dbg_expr) {
+  assert(dbg_expr->GetOpenCL100DebugOpcode() ==
+         OpenCLDebugInfo100DebugExpression);
+  std::unique_ptr<Instruction> deref_expr(dbg_expr->Clone(context()));
+  deref_expr->SetResultId(context()->TakeNextId());
+  deref_expr->InsertOperand(
+      kDebugExpressOperandOperationIndex,
+      {SPV_OPERAND_TYPE_ID, {GetDebugOperationWithDeref()->result_id()}});
```

Comment:
Did you check the order in which the expression should be applied?  How do you know it should be the start of the expression instead of the end?

Reply:
```
struct S {
  bool a;
  char b;
} foo;

// When "%1" is "foo.a"
call void @llvm.dbg.value(metadata i8* %1, metadata !11, metadata !DIExpression(DW_OP_deref, DW_OP_LLVM_fragment, 0, 8)), !dbg !18
```
In the above code, `%1` is `foo.a` and
`DIExpression(DW_OP_deref, DW_OP_LLVM_fragment, 0, 8)` means
1. apply `DW_OP_deref`
2. apply `DW_OP_LLVM_fragment, 0, 8` that mean the member in the offset `0` whose size is `8`

If I change it to `call void @llvm.dbg.value(metadata i8* %1, metadata !11, metadata !DIExpression(DW_OP_LLVM_fragment, 0, 8, DW_OP_deref)), !dbg !18` and generate the assembly code, LLVM prints error
```
invalid expression
!DIExpression(4096, 0, 8, 6)
```
`4096` is the opcode for `DW_OP_LLVM_fragment` and `6` is the opcode for `DW_OP_deref`.

Therefore, in LLVM IR, the order of applying operations of `DIExpression` is from the lower indexed argument to the higher indexed argument.

I quickly checked https://github.com/KhronosGroup/SPIRV-LLVM-Translator/blob/master/lib/SPIRV/SPIRVToLLVMDbgTran.cpp#L781 that is SPIRV-to-LLVM translation for `DebugExpression`.
It uses the same order when converting it to LLVM IR.

In my opinion, we can consider the order must be from the first operation to the last operation.

However, we have to **update the spec for the clarification**.